### PR TITLE
Minimise the need to batch operations in `PlainEditor`

### PR DESCRIPTION
**Pull request by DJMcNab**
The current design of `PlainEditor` and `PlainEditorTransaction` is that if the editor is accessible, the inner layout is always valid and up-to-date (barring a caught panic in the `transact` method). This is not ideal for Masonry, because it forces event handling to recreate the full layout even if the old version could be reused; this potential reuse is especially prevalent in updates driven by Xilem.

This does bring in another change from Masonry, which is `StyleSet`; this is a set of style properties, differentiated by their kind. It can contain at most one font size, line height, etc.
This has some potentially unexpected behaviour with `FontStack`s, as in the other APIs they "combine", but in this format they. However, this combining behaviour is not documented as far as I can see, so I think it's reasonable.
This implemented using a hashmap of discriminants in the enum.
The current API of `Arc<[StyleProperty]>` is horrendously unergonomic.

To reflect this change, I've renamed `PlainEditorTxn` to `PlainEditorDriver`. Suggestions for an alternative name are welcome.

### CODE REVIEW

File: parley/src/layout/editor.rs
Comment (tomcur) [inline parley/src/layout/editor.rs:718]: We may want to add


```rust
debug_assert!(!self.is_composing());
```

here too. This was missed in [xilem#762](https://github.com/linebender/xilem/pull/762).

Comment (DJMcNab) [reply parley/src/layout/editor.rs:718]: Agreed, will do. This gets a bit messy for Xilem; this can happen based on external factors to IME, so I'm not certain how to resolve that...

Comment (tomcur) [reply parley/src/layout/editor.rs]: Good point. We already warn in, e.g., `TextArea` that doing that is likely disruptive if the user is currently focused on the widget. We may be able to just clear compose, set text, and set the compose again, though the compose would likely end up in a somewhat strange location. Alternatively, a more rigorous clear compose, end IME, set text, start IME.

We can discuss this further outside of Parley.

Comment (DJMcNab) [reply parley/src/layout/editor.rs]: Agreed, I don't think this needs to be blocking here.

**Review by jaredoconnell — APPROVED** [review]
I'll defer to those more familiar with the code regarding the design choices, but the changes themselves look good, and I can't find any regressions. Tested with Vello editor and Masonry, Windows and MacOS. Tested accessibility on MacOS.

File: CHANGELOG.md
Comment (tomcur) [inline CHANGELOG.md:28]: We may also want to say support for IME preedit composing text is is introduced into `PlainEditor`.

File: examples/vello_editor/src/text.rs
Comment (nicoburns) [inline examples/vello_editor/src/text.rs:325]: Nit: These could also be `self.driver().clear_compose()`? Seems better not to introduce a closure here if not needed.

Comment (nicoburns) [reply examples/vello_editor/src/text.rs:325]: Even better: I believe you could create one driver at the top of the function. And then everything else is just calling method(s) on that.

File: parley/src/layout/editor.rs
Comment (nicoburns) [inline parley/src/layout/editor.rs:588]: This code seems to have disappeared. Is it no longer considered useful?

**Comment by DJMcNab** [comment]
> 

> The changes seem broadly good. However, there seem to be at least 3 independent changes in this PR:
> 
>  * Change from transaction to dirty tracking
>  * Addition of IME functionality
>  * Change of style representation
> 
> 
> I would prefer to review and land them separately if possible, but I have also tried to add review comments to this PR as-is.

The IME functionality was available to review on its own in https://github.com/linebender/xilem/pull/762, fwiw. I agree that the style change is relatively orthogonal, but both sets of changes were needed to port to `PlainEditor` in Masonry, so seperating them out again would from my perspective have been largely just busywork.

File: parley/src/layout/editor.rs
Comment (DJMcNab) [reply parley/src/layout/editor.rs:733]: Hmm, can you clarify how a user of `PlainEditor` is expected to call the alignment method on `Layout`? As far as I can tell, that method takes exclusive access to self, but the `PlainEditor` doesn't grant exclusive access to the layout.

This comment is largely discussing Masonry semantics, so I'll remove it. Using a sensible available width for alignment is something I would like to do, but it doesn't block this PR.

File: parley/src/layout/editor.rs
Comment (DJMcNab) [reply parley/src/layout/editor.rs:253]: Adding an empty pre-edit is semantically meaningless; having `reset_compose` as a separate method means that it's obvious when you've stopped composing.

File: parley/src/style/styleset.rs
Comment (DJMcNab) [reply parley/src/style/styleset.rs:7]: I guess? Having to implement the conversion back from `TextStyle<'static, Brush>` to an iterator of StyleProperties is quite miserable though, isn't it?

File: parley/src/layout/editor.rs
Comment (DJMcNab) [reply parley/src/layout/editor.rs:899]: Good idea!

File: parley/src/layout/editor.rs
Comment (DJMcNab) [reply parley/src/layout/editor.rs:179]: You are liable to end up in a really janky state.

If you really want us to implement a fallback, then making them a no-op would be the right one. It is, as documented, a logic error to call them, and they can't have any meaningful semantics. Doing an early return would be some quite ugly boilerplate:
```rust
if self.editor.is_composing() {
    if cfg!(debug_assertions) {
         panic!("Can't call most PlainEditor methods whilst composing");
    } else {
         return;
    }
}
```
But if you want us to the check unconditionally, we may as well just make it a full `assert`.

Cancelling compose mode *isn't something we can do* at this level of semantics; to cancel compose mode, you need to tell that platform that you're doing so, and we don't know what the platform is.

File: parley/src/layout/editor.rs
Comment (DJMcNab) [reply parley/src/layout/editor.rs:588]: Restored

**Comment by nicoburns** [comment]
> The IME functionality was available to review on its own in https://github.com/linebender/xilem/pull/762, fwiw.

FWIW, I do not feel competent to review the IME stuff. Previous PR's looked "fine" to me, but apparently had issues on some platforms. Keen to get support merged though!

File: parley/src/style/styleset.rs
Comment (nicoburns) [reply parley/src/style/styleset.rs:7]: I was rather thinking of adding a method to `RangedBuilder` that takes a `TextStyle` directly. `TreeBuilder` already has one, so it could be copied from there.

File: parley/src/layout/editor.rs
Comment (nicoburns) [reply parley/src/layout/editor.rs:179]: > You are liable to end up in a really janky state.
> If you really want us to implement a fallback, then making them a no-op would be the right one.

My main point here being that if debug assertions are disabled, then the current code will lead to any such janky state (because there is no protection other than the debug assertion). Seems to me it should either be an unconditional panic or an early return noop (perhaps also logging a warning). I'd probably prefer the latter as I don't tend to like libraries panicking on me unexpectedly in produciton.

> Doing an early return would be some quite ugly boilerplate:

Perhaps a good case for a macro_rules macro?

File: parley/src/layout/editor.rs
Comment (nicoburns) [reply parley/src/layout/editor.rs]: Does Masonry need to control this? I'm not opposed to adding functionality in for that. I just thought that `PlainEditor` had enough information available to "just do" the right thing for the text input use case.

File: parley/src/layout/editor.rs
Comment (tomcur) [reply parley/src/layout/editor.rs:179]: I'd be happy with it being a noop or assert on release, and an assert in debug mode. The janky state can get as bad as leading to crashes when the preedit byte offsets go out of text buffer bounds.

File: parley/src/layout/editor.rs
Comment (DJMcNab) [reply parley/src/layout/editor.rs]: So, the case I'm thinking about is the scenario where you don't want word wrap, but have a finite available area, and want to be right-aligned (think for example, a spreadsheet app). In that case, your width used for alignment would be the width available, but the width used for layout would be `None`.

Comment (nicoburns) [reply parley/src/layout/editor.rs]: Hmm... if we actually had a way to disable word wrap then that might be a better way to handle that case. But maybe we should add a property to the editor in the short term?

It also brings up the case of a scrollable text field. Might we want to implement that in PlainEditor?

**Comment by DJMcNab** [comment]
I think the "make the operations no-ops" and "TextStyle" changes should both be follow-ups. I believe that the other review comments have been addressed.

File: parley/src/style/styleset.rs
Comment (tomcur) [reply parley/src/style/styleset.rs:7]: I suggested something similar to `TextStyle` elsewhere, where each property is wrapped in an `Option` so we know whether the property is explicitly set or not, but the same concerns around ergonomics remain.

**Review by nicoburns — APPROVED** [review]
Happy to do the remaining bits as followups.

**Pull request merged**
